File: Classification/KNN.R
# K-Nearest Neighbors Classification

# importing dataset
dataset = read.csv('Social_Network_Ads.csv')
dataset = dataset[,3:5]

# splitting dataset into training set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# feature scaling for the independent variables
training_set[,-3] = scale(training_set[,-3])
test_set[,-3] = scale(test_set[-3])

# fitting the KNN classifier and predicting test results
library(class)
y_pred = knn(train = training_set[,-3], test = test_set[,-3],cl = training_set[,3],k=5, prob=TRUE)

# making confusion matrix
cm = table(test_set[,3],y_pred)

# visualising training set results
library(ElemStatLearn)
set =  training_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1,X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = knn(train = training_set[,-3],test = grid_set, cl = training_set[,3], k=5, prob = TRUE)
plot(training_set[,-3], main = 'K-NN Classification - Training Set', 
     xlab = 'age', ylab = 'salary', xlim = range(X1), ylim = range(X2))
contour(X1,X2,matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid ==1, 'springgreen3', 'tomato'))
points(training_set[, -3], pch = 21, bg = ifelse(training_set[,3]==1, 'white', 'black'))


# visualizing test set results
library(ElemStatLearn)
X1 = seq(min(test_set[1]) -1, max(test_set[1]) + 1, by = 0.01)
X2 = seq(min(test_set[2]) -1, max(test_set[2]) + 1, by = 0.01)
grid_set = expand.grid(X1,X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
grid_pred = knn(train = test_set[, 1:2], test = grid_set, cl = test_set[,3],k=5, prob = TRUE)
plot(test_set[,1:2], main = 'K-NN Classfication - Test Set', xlab = 'Age', ylab = 'Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(grid_pred), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(grid_pred ==1, 'springgreen3', 'tomato'))
points(test_set[,1:2], pch = 21, bg = ifelse(y_pred == 1, 'blue', 'white'))

File: Classification/KNN.py
# -*- coding: utf-8 -*-
"""
Created on Fri Nov  3 11:04:11 2017

@author: adhingra
"""
# K-Nearest Neighbors (K-NN)

# importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# importing dataset
dataset = pd.read_csv('Social_Network_Ads.csv')
X = dataset.iloc[:,2:4].values
y = dataset.iloc[:,4].values

# splitting dataset into training set and test set
from sklearn.cross_validation import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=0)

# feature scaling of the indepdendent variables
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# fitting K-Nearest Neighbors Classifier to the training set
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
classifier.fit(X_train,y_train)

# predicting test set results
y_pred = classifier.predict(X_test)

# making confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)

# visualizing training set results
from matplotlib.colors import ListedColormap
X_set,y_set = X_train, y_train
X1,X2 = np.meshgrid(np.arange(start = X_set.min()-1,stop=X_set.max()+1,step=0.01),
                    np.arange(start=X_set.min()-1,stop=X_set.max()+1,step=0.01))
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(),X2.ravel()]).T).reshape(X1.shape),
             cmap=ListedColormap(('red','green')), alpha=0.75)
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j,0],X_set[y_set == j,1],c = ListedColormap(('white','black'))(i), label=j)
plt.title("KNN Classifier - Training set")
plt.legend()
plt.xlabel('age')
plt.ylabel('salary')
plt.show()

# visualizing test set results
from matplotlib.colors import ListedColormap
X_set,y_set = X_test,y_test
X1,X2 = np.meshgrid(np.arange(start=X1.min()-1,stop=X1.max()+1,step=0.01),
                    np.arange(start=X2.min()-1,stop=X2.max()+1,step=0.01))
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(),X2.ravel()]).T).reshape(X1.shape),
             alpha=0.75,cmap=ListedColormap(('red','green')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j,0],X_set[y_set == j,1],
                c=ListedColormap(('yellow','blue'))(i), label =j)
plt.title('KNN Classifier - Test Set')
plt.legend()
plt.xlabel('age')
plt.ylabel('salary')
plt.show()
